application/common/models/EmailQueue.php

```diff
@@ -0,0 +1,219 @@
+                    'targetAttribute' => ['event_log_user_id' => 'id'],
+                ],
+            ],
+            parent::rules()
+        );
+    }
+
+    /**
+     * Attempt to send an email, but on error queue it.
+     * Throws exception if send and queue fail
+     * @param string $toAddress
+     * @param string $subject
+     * @param string $textBody
+     * @param null|string $htmlBody
+     * @param null|string $ccAddress
+     * @param null|integer $eventLogUserId
+     * @param null|string $eventLogTopic
+     * @param null|string $eventLogDetails
+     * @return EmailQueue
+     * @throws \Exception
+     */
+    public static function sendOrQueue(
+        $toAddress,
+        $subject,
+        $textBody,
+        $htmlBody = null,
+        $ccAddress = null,
+        $eventLogUserId = null,
+        $eventLogTopic = null,
+        $eventLogDetails = null
+    ) {
+        $emailQueue = new EmailQueue();
+        $emailQueue->to_address = $toAddress;
+        $emailQueue->subject = $subject;
+        $emailQueue->text_body = $textBody;
+        $emailQueue->html_body = $htmlBody;
+        $emailQueue->cc_address = $ccAddress;
+        $emailQueue->event_log_user_id = $eventLogUserId;
+        $emailQueue->event_log_topic = $eventLogTopic;
+        $emailQueue->event_log_details = $eventLogDetails;
+
+        $emailQueue->send();
+
+        return $emailQueue;
+    }
+
+    /**
+     * Attempt to send email. Returns true on success or throws exception.
+     * @return void
+     * @throws \Exception
+     */
+    public function send()
+    {
+        $log = [
+            'class' => __CLASS__,
+            'action' => 'send',
+            'to' => $this->to_address,
+            'subject' => $this->subject,
+        ];
+
+        /*
+         * Try to send email or throw exception
+         */
+        try {
+            $mailer = $this->getMailer();
+            $sent = $mailer->send();
+            if ($sent === 0 || $sent === false) {
+                throw new \Exception('Unable to send email', 1461011826);
+            }
+
+            /*
+             * Create event log entry if needed
+             */
+            $this->createEventLogEntry();
+
+            /*
+             * Remove entry from queue (if saved to queue) after successful send
+             */
+            $this->removeFromQueue();
+
+            /*
+             * Log success
+             */
+            $log['status'] = 'sent';
+            \Yii::warning($log, 'application');
+
+        } catch (\Exception $e) {
+            /*
+             * Send failed, attempt to queue
+             */
+            $this->attempts_count += 1;
+            $this->last_attempt = Utils::getDatetime();
+            $this->queue($log, $e);
+        }
+    }
+
+    /**
+     * Builds a mailer object from $this and returns it
+     * @return \yii\mail\MessageInterface
+     */
+    private function getMailer()
+    {
+        $mailer = \Yii::$app->mailer->compose();
```

Comment:
For what it's worth, this actually creates a message ([`yii\mail\MessageInterface`](http://www.yiiframework.com/doc-2.0/yii-mail-messageinterface.html)), not a mailer (`yii\mail\MailerInterface`), so it may be worth renaming this from `$mailer` to `$message` (and possibly changing the name of `getMailer()` as well).